nvflare/app_common/abstract/launcher.py

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import threading
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Tuple
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+
+
+class LauncherStatus(str, Enum):
+    SUCCESS = "SUCCESS"
+    FAILED = "FAILED"
+    RUNNING = "RUNNING"
+
+
+class Launcher(ABC):
+    def __init__(self):
+        self.fl_ctx = None
+
+    def initialize(self, fl_ctx: FLContext):
+        self.fl_ctx = fl_ctx
+
+    def finalize(self, fl_ctx: FLContext):
+        pass
+
+    def get_app_dir(self):
+        fl_ctx = self.fl_ctx
+        app_dir = os.path.abspath(fl_ctx.get_engine().get_workspace().get_app_dir(fl_ctx.get_job_id()))
+        return app_dir
+
+    @abstractmethod
+    def launch(self, task_name: str, shareable: Shareable, stop_event: threading.Event):
```

Comment:
I was thinking that users should be able to easily write their own Launcher.
So I try to minimize the "NVFlare" concepts in method signature.
But I still have "shareable" here, let me see if I can further minimize it.

Reply:
thanks for the feedback, I re-evaluate the implementation of "Launcher".
Let's have fl_ctx there since we are already using "shareable" and "abort_signal" and those are NVFlare specific concept.